---
subtitle: "5-scRNAseq initiation for Exp2: the IMs refilling during post deletion"
# bibliography: bib.bib
link-citations: yes
date: "`r format (Sys.time(), format='%Y-%m-%d %H:%M:%S %z')`"
output: 
  pdf_document:
    pandoc_args: 
      - '../common.yaml'
      - --listings
    includes: 
      in_header: '../preamble.tex'
    toc: true
    number_sections: true
    fig_width: 7
    fig_height: 6
    fig_caption: true
    df_print: tibble
---

```{r include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

\newpage

# Description

Lung interstitium macrophages (IMs) are non-alveolar resident tissue macrophages which contribute to the lung homeostasis. These cells were reported to be heterogeneous by our group and other teams, which contains two main distinct subpopulations: CD206+ IMs and CD206- IMs. However, the exact origin of IMs and the transcriptional programs that regulate IM differentiation remains unclear. In recent report, we analyzed the refilled IMs in the course of time after induced IM depletion with single-cell RNA sequencing (10X Genomics Chromium) and bulk RNA sequencing. 

The lung IMs and monocytes from the mice at 12 hours (DT12h), 24 hours (DT24h) and 48 hours (DT48h) after diphtheria toxin (DT)-induced IM depletion were analyzed and compared using single-cell RNA sequencing. A subpopulation was found to be a transit differentiating cells from monocytes to IMs. Transcription factor activity analysis and trajectory showed cMAF and MAFb transcription factors played important roles in monocyte-IM differentiation. 

Over design of experiment: 

The lung IMs and monocytes were sorted from three groups of mice: 
Group “DT12h” contains 5 IM-DTR mice which were treated with 50 ng diphtheria toxin (DT) 12 hours before sacrifice. 
Group “DT24h” contains 5 IM-DTR mice which were treated with 50 ng DT 24 hours before sacrifice. 
Group “DT48h” contains 5 IM-DTR mice which were treated with 50 ng DT 48 hours before sacrifice. 

Cells from each group were barcoded with different Biolegend anti-mouse Hastags before being pooled for library construction. Processed data then were analyzed with Bioconductor and Seurat package. The cells from different time points were demultiplexed with the barcode detected in each cell. 

# Demultiplexing (de-hashtag)

## Load packages and data

```{r load_pacakges}
library(Seurat)
```

```{r include=FALSE}
# Load in the UMI matrix (from CR)

immune.umis <- Read10X("../../../IM_DTR_exp2/counts/scRNAseq/chromium/CD45Plus_NGS21-S229/outs/filtered_feature_bc_matrix")

# For generating a hashtag count matrix from FASTQ files, please refer to https://github.com/Hoohm/CITE-seq-Count.
# Load in the HTO count matrix
immune.htos <- Read10X("../../../IM_DTR_exp2/counts/scRNAseq/HTO/CD45Plus_NGS21-S229/read_count", gene.column=1)
```

```{r read_Data}
# change cell names to "-1" mode. 
colnames(immune.htos) <- paste0(colnames(immune.htos), "-1")

# remove the unmapped as it's not a barcode. It's the last row. 
immune.htos <- immune.htos[-nrow(immune.htos), ]

# remove the "HT5_CTL-CTTTGTCTTTGTGAG" as it's not added to the sample. 
immune.htos <- immune.htos[rownames(immune.htos) != "HT5_CTL-CTTTGTCTTTGTGAG", ]

# Select cell barcodes detected by both RNA and HTO
# In the example datasets we have already filtered the cells for you, but perform this step for clarity.
joint.bcs <- intersect(colnames(immune.umis), colnames(immune.htos))

# Subset RNA and HTO counts by joint cell barcodes
immune.umis <- immune.umis[, joint.bcs]
immune.htos <- as.matrix(immune.htos[, joint.bcs])


# Confirm that the HTO have the correct names
rownames(immune.htos)
```

## Setup Seurat object and add in the HTO data

```{r hashtag_setup}
# Setup Seurat object
immune.hashtag <- CreateSeuratObject(counts = immune.umis)

# Normalize RNA data with log normalization
immune.hashtag <- NormalizeData(immune.hashtag)
# Find and scale variable features
immune.hashtag <- FindVariableFeatures(immune.hashtag, selection.method = 'mean.var.plot')
immune.hashtag <- ScaleData(immune.hashtag, features = VariableFeatures(immune.hashtag))
```

## Adding HTO data as an independent assay

```{r hto_assay}
# Add HTO data as a new assay independent from RNA
immune.hashtag[['HTO']] <- CreateAssayObject(counts = immune.htos)
# Normalize HTO data, here we use centered log-ratio (CLR) transformation
immune.hashtag <- NormalizeData(immune.hashtag, assay = 'HTO', normalization.method = 'CLR')
```

## Demultiplex cells based on HTO enrichment

Here we use the Seurat function `HTODemux()` to assign single cells back to their sample origins.

```{r hashtag_demux}
# Here we are using the default settings
immune.hashtag <- HTODemux(immune.hashtag, assay = "HTO", positive.quantile = 0.99)
```

## Visualize demultiplexing results

Output from running `HTODemux()` is saved in the object metadata. We can visualize how many cells are classified as singlets, doublets and negative/ambiguous cells.

```{r demux_summary}
# Global classification results
table(immune.hashtag$HTO_classification.global)
```

Visualize enrichment for selected HTOs with ridge plots

```{r hashtag_ridge, fig.width=16, fig.height=10}
# Group cells based on the max HTO signal
Idents(immune.hashtag) <- 'HTO_maxID'
RidgePlot(immune.hashtag, assay = 'HTO', features = rownames(immune.hashtag[['HTO']]), ncol = 2)
```

Visualize pairs of HTO signals to confirm mutual exclusivity in singlets

```{r hashtag_scatter1, fig.height=8, fig.width=9}
FeatureScatter(immune.hashtag, feature1 = 'HT6-12hDT-TATGCTGCCACGGTA', feature2 = 'HT7-24hDT-GAGTCTGCCAGTATC')
```

```{r hashtag_scatter2, fig.height=8, fig.width=9}
FeatureScatter(immune.hashtag, feature1 = 'HT7-24hDT-GAGTCTGCCAGTATC', feature2 = 'HT8-48hDT-TATAGAACGCCAGGC')
```

Compare number of UMIs for singlets, doublets and negative cells
```{r hashtag_vln, fig.width=10}
Idents(immune.hashtag) <- 'HTO_classification.global'
VlnPlot(immune.hashtag, features = 'nCount_RNA', pt.size = 0.1, log = TRUE)
```

Generate a two dimensional tSNE embedding for HTOs.Here we are grouping cells by singlets and doublets for simplicity.

```{r hashtag_sub_tsne, fig.width=9}
#First, we will remove negative cells from the object
immune.hashtag.subset <- subset(immune.hashtag, idents = 'Negative', invert = TRUE)

# Calculate a tSNE embedding of the HTO data
DefaultAssay(immune.hashtag.subset) <- "HTO"
immune.hashtag.subset <- ScaleData(immune.hashtag.subset, features = rownames(immune.hashtag.subset), verbose = FALSE)
immune.hashtag.subset <- RunPCA(immune.hashtag.subset, features = rownames(immune.hashtag.subset), approx = FALSE)
immune.hashtag.subset <- RunTSNE(immune.hashtag.subset, dims = 1:8, perplexity = 100, check_duplicates = FALSE)

DimPlot(immune.hashtag.subset)
```

Create an HTO heatmap

```{r hashtag_heatmap, fig.width=12}
HTOHeatmap(immune.hashtag, assay = 'HTO', ncells = 5000)
```

Cluster and visualize cells using the usual scRNA-seq workflow, and examine for the potential presence of batch effects.

```{r hastag_cluster, message=FALSE, warning=FALSE}
# Extract the singlets
immune.singlet <- subset(immune.hashtag, idents = 'Singlet')

# Select the top 1000 most variable features
immune.singlet <- FindVariableFeatures(immune.singlet, selection.method = 'mean.var.plot')

# Scaling RNA data, we only scale the variable features here for efficiency
immune.singlet <- ScaleData(immune.singlet, features = VariableFeatures(immune.singlet))

# Run PCA
immune.singlet <- RunPCA(immune.singlet, features = VariableFeatures(immune.singlet))
```

```{r hashtag_tsne, fig.width=9}
# We select the top 10 PCs for clustering and tSNE based on PCElbowPlot
immune.singlet <- FindNeighbors(immune.singlet, reduction = 'pca', dims = 1:10)
immune.singlet <- FindClusters(immune.singlet, resolution = 0.6, verbose = FALSE)
immune.singlet <- RunTSNE(immune.singlet, reduction = 'pca', dims = 1:10)

# Projecting singlet identities on TSNE visualization
DimPlot(immune.singlet, group.by = "HTO_classification")
```

save to seurat object: 
```{r eval=FALSE}
saveRDS(immune.singlet, file = "./immune.demultiplexed.seuratObject.rds")
```

# QC of pooled sample

Load data: 
```{r message=FALSE}
# load package and data
library(ggpubr)
source("~/Desktop/velocyto/Script/seurat.setup.R")
dir.10x <- "/mnt/Data/Single-cell_Analysis/Projects/IPL/IM_DTR/IM_DTR_exp2/counts/scRNAseq/chromium"
```

# QC

## Sample: CD45Plus_NGS21-S229
```{r message=FALSE, warning=FALSE}

CD45Plus_NGS21_S229 <- seurat.setup(path.10x = file.path(dir.10x, "CD45Plus_NGS21-S229/outs/filtered_feature_bc_matrix/") , project = "IM-DTR", dimensionality = 1:20, mt.percentage = 20, human = FALSE)
```

```{r fig.width=12, fig.height=10}
ggarrange(CD45Plus_NGS21_S229$plots$feature_vln, CD45Plus_NGS21_S229$plots$RNA_mt.pct_scatter, CD45Plus_NGS21_S229$plots$JackStrawPlot, CD45Plus_NGS21_S229$plots$ElbowPlot, ncol = 2, nrow = 2)
```

```{r fig.width=12, fig.height=10}
ggarrange(CD45Plus_NGS21_S229$plots$variable_features, CD45Plus_NGS21_S229$plots$PCA_plot, CD45Plus_NGS21_S229$plots$UMAP_plot, CD45Plus_NGS21_S229$plots$TSNE_plot, ncol = 2, nrow = 2)
```

```{r eval=FALSE}
# save data for next use: 
saveRDS(CD45Plus_NGS21_S229$seuratObject, file = "./CD45Plus_NGS21_S229.seuratObject.rds")
```

# Identify the cells for each samples

```{r include=FALSE}
CD45Plus_NGS21_S229.seuratObject <- CD45Plus_NGS21_S229$seuratObject
immune.demultiplexed.seuratObject <- immune.singlet
```

Assign HTO calling to the Seurat object
```{r}
common.immune <- intersect(colnames(CD45Plus_NGS21_S229.seuratObject), colnames(immune.demultiplexed.seuratObject))
CD45Plus_NGS21_S229.seuratObject <- subset(CD45Plus_NGS21_S229.seuratObject, cells = common.immune)
```


```{r}
# make intersect between hto and rna cells:
hto.immune <- immune.demultiplexed.seuratObject@meta.data$HTO_classification
names(hto.immune) <- colnames(immune.demultiplexed.seuratObject)

# remove the sequence chars
hto.immune <- sub(hto.immune, pattern = "-[C,T,G,A]{0,15}$", replacement = "")

# assign to seurat object: 
CD45Plus_NGS21_S229.seuratObject$treatment <- hto.immune[colnames(CD45Plus_NGS21_S229.seuratObject)]
```
    
## Build metadata table

```{r}
# cell type
CD45Plus_NGS21_S229.seuratObject$cell.type0 <- "CD45+"
```

save individual samples for other use:
```{r}
obj <- CD45Plus_NGS21_S229.seuratObject
for (i in unique(obj$treatment)) {
  obj.sub <- subset(obj, subset = treatment == i)
  file.name <- paste("CD45Plus_NGS21_S229", i, "seuratObject.rds", sep = ".")
  # saveRDS(object = obj.sub, file = file.path(".", file.name)) # Here we can also save the individual object.
}
```

## Data processing for CD45Plus_NGS21_S229

```{r}
CD45Plus_NGS21_S229.seuratObject <- NormalizeData(CD45Plus_NGS21_S229.seuratObject)
CD45Plus_NGS21_S229.seuratObject <- FindVariableFeatures(CD45Plus_NGS21_S229.seuratObject, 
                                                         selection.method = "vst", nfeatures = 2000)
CD45Plus_NGS21_S229.seuratObject <- ScaleData(CD45Plus_NGS21_S229.seuratObject, 
                                              features = rownames(CD45Plus_NGS21_S229.seuratObject))
```

## Linear dimension reduction
```{r}
CD45Plus_NGS21_S229.seuratObject <- RunPCA(CD45Plus_NGS21_S229.seuratObject, 
                                           features = VariableFeatures(object = CD45Plus_NGS21_S229.seuratObject))
CD45Plus_NGS21_S229.seuratObject <- JackStraw(CD45Plus_NGS21_S229.seuratObject, num.replicate = 100)
CD45Plus_NGS21_S229.seuratObject <- ScoreJackStraw(CD45Plus_NGS21_S229.seuratObject, dims = 1:20)
JackStrawPlot(CD45Plus_NGS21_S229.seuratObject, dims = 1:20)
```

```{r}
CD45Plus_NGS21_S229.seuratObject <- FindNeighbors(CD45Plus_NGS21_S229.seuratObject, dims = 1:30)
CD45Plus_NGS21_S229.seuratObject <- FindClusters(CD45Plus_NGS21_S229.seuratObject, resolution = 0.5)
```
```{r}
CD45Plus_NGS21_S229.seuratObject <- RunUMAP(CD45Plus_NGS21_S229.seuratObject, dims = 1:30)
CD45Plus_NGS21_S229.seuratObject <- RunUMAP(CD45Plus_NGS21_S229.seuratObject, dims = 1:30)
```

```{r}
DimPlot(CD45Plus_NGS21_S229.seuratObject)
```

# Identify cell types

## SingleR analysis

```{r message=FALSE, warning=FALSE}
source("../R/seurat2singleR.R")
results.singleR <- seurat2singleR(CD45Plus_NGS21_S229.seuratObject, ref = "ImmGenData")
```

```{r eval=FALSE}
saveRDS(results.singleR, file = "./CD45Plus_NGS21_S229.ImmGenData.singleR.Rds")
saveRDS(CD45Plus_NGS21_S229.seuratObject, file = "./CD45Plus_NGS21_S229.seuratObject.rds")
```


## Marker expression check

```{r message=FALSE}
results <- CD45Plus_NGS21_S229.seuratObject
DimPlot(results, reduction = "umap", label = TRUE)
```

```{r fig.width=10, fig.height=10}
DefaultAssay(results) <- "RNA"
FeaturePlot(results, features = c("Fcgr1", "H2-Eb1", "Cd74", "Mrc1", "Folr2", "Cd163", "Ace", "Fcgr4"), reduction = "umap") 
```




## Combine marker expression with SingleR results 

```{r fig.width=8, fig.height=7}
results$singleR.celltype <- results.singleR$labels

DimPlot(results, group.by = "singleR.celltype", reduction = "umap", label = T) + ggtitle("DatabaseImmuneCellExpressionData - SingleR identity")
```

```{r fig.width=8, fig.height=7}
DimPlot(results, group.by = "singleR.celltype", reduction = "tsne", label = T) + ggtitle("DatabaseImmuneCellExpressionData - SingleR identity")
```

Cell numbers of each type: 

```{r}
table(results$singleR.celltype)
```

Split by treatment: contaminated cells came from the same sample?

```{r fig.height=8, fig.width=16}
DimPlot(results, group.by = "singleR.celltype", reduction = "umap", label = T, split.by = "treatment") + ggtitle("results - SingleR identity (split by treatment)")
```

## Remove contaminated cell types other than monocytes/macrophages

```{r}
results <- subset(results, subset = singleR.celltype %in% c("B cells", "Endothelial cells", "Eosinophils", "Fibroblasts", "ILC", "Neutrophils", "NK cells", "NKT", "Stem cells", "T cells", "Tgd"), invert = TRUE)
```
*Here we keep some closely-related cell types, like DC, microglia etc.*

# Phenotypic characterization of monocytes/macrophages

## Data processing after filtering:

Remove old snn: 
```{r}
results$RNA_snn_res.0.5 <- NULL
```

```{r message=FALSE, warning=FALSE}
DefaultAssay(results) <- "RNA"
results <- NormalizeData(results)
results <- FindVariableFeatures(results, selection.method = "vst", nfeatures = 2000)
results <- ScaleData(results, features = rownames(results))
results <- RunPCA(results, features = VariableFeatures(results))
```

Non-linear reduction: 
*Let's choose dim = 1:10, showing very decent bridge between monocytes and macrophages.*

```{r message=FALSE, warning=FALSE}
results <- RunTSNE(results, dims = 1:10)
results <- RunUMAP(results, dims = 1:10)
```

```{r fig.width=10, fig.height=4}
DimPlot(results, reduction = "umap", split.by = "treatment")
```


## Clustering of cells 

```{r message=FALSE, warning=FALSE}
results <- FindNeighbors(results, dims = 1:10)
results <- FindClusters(results, resolution = 0.14) 
```

```{r fig.width=12, fig.height=5}
Idents(results) <- "RNA_snn_res.0.14"
DimPlot(results, label = TRUE, split.by = "treatment")
```

Distribution of clusters across samples: 

```{r fig.width=3, fig.height=4}
source("../R/SeuratFreqTable.R")
freq.celltype.list <- list(
  `12hDT` = Seurat2CellFreqTable(subset(results, subset = treatment == "HT6-12hDT"), slotName = "RNA_snn_res.0.14"),  
  `24hDT` = Seurat2CellFreqTable(subset(results, subset = treatment == "HT7-24hDT"), slotName = "RNA_snn_res.0.14"),
  `48hDT` = Seurat2CellFreqTable(subset(results, subset = treatment == "HT8-48hDT"), slotName = "RNA_snn_res.0.14") 
)

source("../R/barChart.R")
barChart(freq.celltype.list) + labs(fill = "Cluster")
```

## Population characterization

```{r fig.width=8, fig.height=15}
library(dplyr)
all_cluster.markers <- FindAllMarkers(results)
top20 <- all_cluster.markers %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
DoHeatmap(results, features = top20$gene) + NoLegend()
```

Expression of relevant genes 

```{r fig.height=16, fig.width=16}
VlnPlot(results, features = c("Ccr2", "Plac8" ,"Ccr2", "Spn", 
                              "Fcgr4", "Ace", "Cxcr4", 
                              "Mrc1", "Cd63", "Apoe", 
                              "H2-DMb2", "H2-Eb1", "Cd74", "Cd209a", 
                              "Ccnd1","Mki67", "Top2a", "Ccl2", 
                              "Irf7", "Cxcl10", "Ifit1", "Ifit2", "Ifi47", "Itgax") , 
        # split.by = "treatment", 
        ncol = 4)
```

*Here's the cell type annotation: *
cluster 0: Classical Mono
cluster 1: Patrolling Mono
cluster 2: MHCII IM
cluster 3: chemoattractant
cluster 4: CD206 IM
cluster 5: Transit

```{r}
levels(results)
```

```{r fig.width=12, fig.height=5}
results$cell.type2 <- factor(Idents(results), labels = c("Classical Mono", "Patrolling Mono", "MHCII IM", "chemoattractant", "CD206 IM", "Transit"))
Idents(results) <- "cell.type2"
DimPlot(results, label = TRUE, split.by = "treatment")
```


## Cell cycle analysis

```{r}
library(cowplot)
data("geneinfo_human", package = "nichenetr")
s.genes <- nichenetr::convert_human_to_mouse_symbols(cc.genes.updated.2019$s.genes)
g2m.genes <- nichenetr::convert_human_to_mouse_symbols(cc.genes.updated.2019$g2m.genes)
results <- CellCycleScoring(results, s.features = s.genes, g2m.features = g2m.genes, set.ident = FALSE)
DimPlot(results, group.by = "Phase")
```
```{r fig.height=5, fig.width=12}
DimPlot(results, group.by = "Phase", reduction = "umap", split.by = "treatment")

```

cluster 0: Classical Mono
cluster 1: Patrolling Mono
cluster 2: MHCII IM
cluster 3: chemoattractant
cluster 4: CD206 IM
cluster 5: Transit

```{r}
freq.celltype.list <- list(
  `Classical Mono` = Seurat2CellFreqTable(subset(results, ident = "Classical Mono"), slotName = "Phase"), 
  `Patrolling Mono` = Seurat2CellFreqTable(subset(results, ident = "Patrolling Mono"), slotName = "Phase"), 
  `MHCII IM` = Seurat2CellFreqTable(subset(results, ident = "MHCII IM"), slotName = "Phase"), 
  `chemoattractant` = Seurat2CellFreqTable(subset(results, ident = "chemoattractant"), slotName = "Phase"), 
  `CD206 IM` = Seurat2CellFreqTable(subset(results, ident = "CD206 IM"), slotName = "Phase"), 
  `Transit` = Seurat2CellFreqTable(subset(results, ident = "Transit"), slotName = "Phase")
)
barChart(freq.celltype.list) + labs(fill = "Cell-cycle phase")
```

## Apoptosis rate by mitochodrial genes

```{r fig.height=5, fig.width=12}
FeaturePlot(results, features = "percent.mt", reduction = "umap", split.by = "treatment")

```

```{r eval=FALSE}
saveRDS(results, file = "./immune.cellType_filtered.withSingleR.seuratObject.Rds")
```


# Session information

```{r}
sessionInfo()
```

# References
